### PR TITLE
Async to_gpu

### LINKED ISSUE #886: Asynchronous Host-to-Device copy
Current CuPy does not have a way to allocate page-locked host memory to enable asynchronous HtoD copy. There are e.g. two options to support asynchronous HtoD copy:

1. Allocate page-locked host memory and copy the source buffer to it implicitly in the MemoryPointer.copy_from_host_async() method. It adds an overhead on copying, which might be problematic.
2. Provide a way to create a numpy array on page-locked host memory. I do not investigate the feasibility yet (maybe it can be done with using np.frombuffer and adding a destructor?).

### PR DESCRIPTION
**Pull request by okuta**
fix #886

### CODE REVIEW

File: chainer/cuda.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ def __eq__(self, other):
     def __ne__(self, other):
         return not (self == other)
 
+    @property
+    def id(self):
+        return -1
```
Comment (beam2d) [inline chainer/cuda.py]: Is it better to make it a class attribute?

File: chainer/cuda.py [hunk 2]
```diff
@@ -190,15 +195,33 @@ def to_gpu(array, device=None, stream=None):
...
+                # cpu to gpu
+                src = array.copy(order='C')
+                ret.set(src, stream)
+            else:
+                # gpu to gpu
+                with array_dev:
+                    src = array.copy()
+                ret.data.copy_from_device_async(src.data, src.nbytes, stream)
+
+            # to hold a reference to the end of the asynchronously memcpy
```
Comment (beam2d) [inline chainer/cuda.py]: to the end -> until the end
asynchronously -> asynchronous

**Comment by beam2d** [comment]
Sorry for being late. LGTM except above comments!

**Comment by okuta** [comment]
@beam2d I fixed.

**Comment by beam2d** [comment]
LGTM!

**Pull request merged**